lib/archive/archive.go

```diff
@@ -109,13 +109,42 @@ func Extract(r io.Reader, dir string) error {
 			return trace.Wrap(err)
 		}
 
-		if err := extractFile(tarball, header, dir); err != nil {
+		if err := extractFile(tarball, header, dir, header.Name); err != nil {
 			return trace.Wrap(err)
 		}
 	}
 	return nil
 }
 
+// ExtractWithPrefix extracts the contents of the specified tarball under directory dir.
+// Only the files/directories found in tarDirPrefix in the tarball matching patterns are extracted.
+// Note, files from tarDir are written directly to dir omitting tarDir.
+// The resulting files and directories are created using the current user context.
+func ExtractWithPrefix(r io.Reader, dir, tarDirPrefix string) error {
+	err := TarGlobWithPrefix(tar.NewReader(r), tarDirPrefix, func(match *tar.Header, r *tar.Reader) error {
+		relpath, err := filepath.Rel(tarDirPrefix, match.Name)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		// Security: ensure tar doesn't refer to file paths outside the directory
+		err = validPath(relpath, dir, "invalid file path")
```

Comment:
I think my only worry is, that this replicates other sanitization code, which leaves some risk if something new is discovered missing in the tar sanitization, that the semi-duplicate code get's missed. I see why it was separated out, that's just my only thought on the risks of doing it this way.

Reply:
I updated to keep using `SanitizeTarPath` as is.